### PR TITLE
Article on Docker, Nix and reprodicibility.

### PR DESCRIPTION
**Pull request by teh**

### CODE REVIEW

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+the GCC compiler - are invisible and taken for granted, but they are
+still dependencies.
+
+Recently I read a few blog posts where people got excited about using
+Docker for reproducibility, and that is a scary idea. What follows is
+a fairly common part of a `Dockerfile`, in this case for the
+[Jupyter notebook](https://github.com/jupyter/notebook/blob/master/Dockerfile):
+
+```
+RUN apt-get update && apt-get install -y language-pack-en
```
Comment (Keats) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: this is a pretty bad example of the dockerfile, I would copy https://github.com/jupyter/notebook/blob/master/Dockerfile#L22-L42

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+In practice things are more complicated: Modern compilers are often
+non-deterministic, especially when building in parallel. I.e. building
+the exact same file twice does not necessarily result in the same
+object code. Many distributions are spending considerable resources on
+improving this situation and Nix
+[is no exception](https://github.com/NixOS/nixpkgs/pull/2281).
+
+Another source of variation is that linear algebra libraries like
+OpenBLAS have specialised code for most CPU architectures, selecting
+the fasted option at runtime.
```
Comment (Keats) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: fastest

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+Instead I want to focus on how researchers can run published source
+code on their own machines and arrive at the same results.
+
+<!-- PELICAN_END_SUMMARY -->
+
+## The issue with Docker
+
+Unless you are publishing
+[MIX assembly](http://en.wikipedia.org/wiki/MIX) all modern code comes
+with a complex set of dependencies. Many of those dependencies - like
```
Comment (Keats) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: or binaries statically linked ! maybe the answer to research is go !

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+the GCC compiler - are invisible and taken for granted, but they are
+still dependencies.
+
+Recently I read a few blog posts where people got excited about using
```
Comment (Keats) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: give me some examples of blog posts

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
+Title: Why Docker is not the answer to reproducible research, and why Nix may be
+Date: 2015-05-17
+Short_summary: We look at how to do properly reproducible research, and why using Docker doesn't work.
+Category: Dev
```
Comment (Keats) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: Gautier used the infrastructure category

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md (hunk 3 above)
Comment (teh) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: Binaries don't tell you where the source came from though. With nix you know every single source package and its hash.

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md (hunk 1 above)
Comment (teh) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: Done though clipped it a bit.

**Comment by teh** [comment]
Updated w comments & rephrased a few sentences to be slightly less aggressive in tone.

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md (hunk 5 above)
Comment (teh) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: done

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md (hunk 4 above)
Comment (teh) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: done

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md (hunk 2 above)
Comment (teh) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: done

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+So the outcome of this `RUN` statement depends on:
+
+* The time the command was run,
+* the build environment of the package uploader,
+* the base image in the Docker registry,
+* and potentially on the package mirror used.
+
+Now multiply this behaviour by many packages and their dependencies.
+
+Reproducing your results even three years from now seems like an
```
Comment (Gautier) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: Reproducing your results, even three years from now, seems

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+the fastest option at runtime.
+
+
+## Summary
+
+While Docker is tempting for shipping research I think it's the wrong
+answer in its current incarnation. Between the systems that are usable
+right now it looks like Nix gives us the best shot at
+reproducibility. Nix still has a long way to go but it is build on a
+sound foundation, not on shifting sands like current Docker.
```
Comment (Gautier) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: just enough controversy to get a discussion going :)

File: content/articles/reproducible-research-or-why-docker-is-not-the-answer.md [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+    md5 = "599881281d45db4ce9adb2d75458391e";
+  };
+
+  buildInputs = [ boost cmake gdal libgeotiff libtiff LASzip];
+
+  # ... Removed some metadata to avoid obscuring the point
+}
+```
+
+Read that as a function taking parameters like the `boost`, `cmake`,
```
Comment (Gautier) [inline content/articles/reproducible-research-or-why-docker-is-not-the-answer.md]: this is related to the small grain of salt I'd add to this article: When we write our default.nix with "with import <nixpkgs> {};" it creates a dependency on the specific version of the nixpkgs installed. We've been bitten by this a couple of times already.

**Pull request merged**